### PR TITLE
Tooltips: Add more BP and stat modifiers

### PR DESCRIPTION
**Pull request by MacChaeger**
Things that I wasn't sure about adding:
- STAB and Adaptability
- Weather and Terrain

### CODE REVIEW

**Comment by Zarel** [comment]
Still failing test

**Comment by Zarel** [comment]
(You have mixed tabs and spaces)

File: js/client-battle-tooltips.js [hunk 1]
```diff
@@ -1152,6 +1165,47 @@ var BattleTooltips = (function () {
...
+
+		var allyActive = pokemon.side.active;
+		var auraBoosted = '';
+		var auraBroken = false;
+		if (move.category !== 'Status' && allyActive.length > 1) {
+			for (var i = 0; i < allyActive.length; i++) {
+				var ally = allyActive[i];
+				if (!ally || ally.fainted) continue;
+				if (!auraBoosted && (ally.ability === 'Fairy Aura' && move.type === 'Fairy') ||
+				    		    (ally.ability === 'Dark Aura' && move.type === 'Dark')) {
```
Comment (Zarel) [inline js/client-battle-tooltips.js]: We definitely don't support this indent+space thing.

File: js/client-battle-tooltips.js [hunk 2]
```diff
@@ -1152,6 +1165,47 @@ var BattleTooltips = (function () {
...
+				}
+			}
+		}
+		var foeActive = this.battle.yourSide.active;
+		if (!auraBoosted && move.category !== 'Status' && (move.type === 'Fairy' || move.type === 'Dark')) {
+			for (var i = 0; i < foeActive.length; i++) {
+				var foe = foeActive[i];
+				if (!foe || foe.fainted) continue;
+				if (!auraBoosted && (foe.ability === 'Fairy Aura' && move.type === 'Fairy') ||
+				    		    (foe.ability === 'Dark Aura' && move.type === 'Dark')) {
```
Comment (Zarel) [inline js/client-battle-tooltips.js]: ditto

File: js/client-battle-tooltips.js (hunk 1 above)
Comment (Slayer95) [reply js/client-battle-tooltips.js]: Also note that && has higher operator precedence than ||.

File: js/client-battle-tooltips.js [hunk 3]
```diff
@@ -1152,6 +1165,47 @@ var BattleTooltips = (function () {
...
+				}
+			}
+		}
+		var foeActive = this.battle.yourSide.active;
+		if (!auraBoosted && move.category !== 'Status' && (move.type === 'Fairy' || move.type === 'Dark')) {
+			for (var i = 0; i < foeActive.length; i++) {
+				var foe = foeActive[i];
+				if (!foe || foe.fainted) continue;
+				if (!auraBoosted && (foe.ability === 'Fairy Aura' && move.type === 'Fairy') ||
+						    (foe.ability === 'Dark Aura' && move.type === 'Dark')) {
```
Comment (Zarel) [inline js/client-battle-tooltips.js]: This indentation is still dumb even if `eslint` doesn't complain about it. This is literally only designed for 8-space tabs, which we don't actually use.

File: js/client-battle-tooltips.js [hunk 4]
```diff
@@ -1152,6 +1165,47 @@ var BattleTooltips = (function () {
...
+
+		var allyActive = pokemon.side.active;
+		var auraBoosted = '';
+		var auraBroken = false;
+		if (move.category !== 'Status' && allyActive.length > 1) {
+			for (var i = 0; i < allyActive.length; i++) {
+				var ally = allyActive[i];
+				if (!ally || ally.fainted) continue;
+				if (!auraBoosted && (ally.ability === 'Fairy Aura' && move.type === 'Fairy') ||
+							(ally.ability === 'Dark Aura' && move.type === 'Dark')) {
```
Comment (Zarel) [inline js/client-battle-tooltips.js]: I actually don't know what the convention is here. I think Cathy wanted to do +2 tabs, but I think consensus is on +1 tab?

Comment (Zarel) [reply js/client-battle-tooltips.js]: +3 tabs is right out, though.

Comment (Slayer95) [reply js/client-battle-tooltips.js]: Might as well delete the ``!auraBoosted`` check and there is no longer a debate to have.
That would also fix the ``&&`` vs ``||`` precedence issue.

**Comment by Slayer95** [comment]
@MacChaeger , my comments re: precedence refer to:

The following code
```js
if (!auraBoosted && (ally.ability === 'Fairy Aura' && move.type === 'Fairy') || (ally.ability === 'Dark Aura' && move.type === 'Dark')) {
// Do Stuff
}
```

is parsed as

```js
if (!auraBoosted && (ally.ability === 'Fairy Aura' && move.type === 'Fairy')) {
// Do Stuff 
} else if (ally.ability === 'Dark Aura' && move.type === 'Dark') {
// Do Stuff
}
```

rather than

```js
if (!auraBoosted) {
if ((ally.ability === 'Fairy Aura' && move.type === 'Fairy') || (ally.ability === 'Dark Aura' && move.type === 'Dark')) {
// Do Stuff
}
}
```

**Comment by MacChaeger** [comment]
OH! I get it now. I forgot a set of parentheses.

File: js/client-battle-tooltips.js [hunk 5]
```diff
@@ -1152,6 +1165,45 @@ var BattleTooltips = (function () {
...
 		}
+
+		var allyActive = pokemon.side.active;
+		var auraBoosted = '';
+		var auraBroken = false;
+		if (move.category !== 'Status' && allyActive.length > 1) {
+			for (var i = 0; i < allyActive.length; i++) {
+				var ally = allyActive[i];
+				if (!ally || ally.fainted) continue;
+				if (!auraBoosted && ((ally.ability === 'Fairy Aura' && move.type === 'Fairy') || (ally.ability === 'Dark Aura' && move.type === 'Dark'))) {
```
Comment (Zarel) [inline js/client-battle-tooltips.js]: This is way too complicated of a condition and would probably benefit from being simplified in some way.

My first thought is that `!auraBoosted &&` isn't really a necessary condition. It's a micro-optimization that makes the code less readable so I'd probably just leave it off entirely.

Comment (Zarel) [reply js/client-battle-tooltips.js]: My second thought is just to split up the `||` into a chain of `switch`es...

```
if (ally.ability === 'Fairy Aura') {
  if (move.type === 'Fairy') auraBoosted = 'Fairy Aura';
} else if (ally.ability === 'Dark Aura') {
  if (move.type === 'Dark') auraBoosted = 'Dark Aura';
} else if (ally.ability === 'Aura Break') {
  auraBroken = true;
} else if (ally.ability === 'Battery') {
  if (ally !== pokemon && move.category === 'Special') {
    // ...
```

File: js/client-battle-tooltips.js [hunk 6]
```diff
@@ -1206,6 +1271,10 @@ var BattleTooltips = (function () {
 		// Type-enhancing items
 		if (itemTypes[item.name] === moveType) return this.battle.gen < 4 ? 1.1 : 1.2;
 
+		// Pokemon-specific items
+		if (item.name === 'Soul Dew' && this.battle.gen < 7) return 0;
+		if (orbTypes[item.name] === moveType || moveType === 'Dragon') return 1.2;
```
Comment (Zarel) [inline js/client-battle-tooltips.js]: If I'm reading this correctly, this will make all Dragon-type moves 1.2x damage?

**Comment by MacChaeger** [comment]
OK, with that, what do you think of showing STAB, weather and terrain? Also, I just found out Victory Star's accuracy boost isn't showing up in the tooltips right now and I can't figure out why.

File: js/client-battle-tooltips.js [hunk 7]
```diff
@@ -1206,6 +1278,10 @@ var BattleTooltips = (function () {
 		// Type-enhancing items
 		if (itemTypes[item.name] === moveType) return this.battle.gen < 4 ? 1.1 : 1.2;
 
+		// Pokemon-specific items
+		if (item.name === 'Soul Dew' && this.battle.gen < 7) return 0;
+		if (orbUsers[pokemonData.baseSpecies] === item.name && orbTypes[item.name] === moveType || moveType === 'Dragon') return 1.2;
```
Comment (Zarel) [inline js/client-battle-tooltips.js]: This still shows 1.2x damage for a Flygon using Outrage.

You need parens.

**Pull request merged**